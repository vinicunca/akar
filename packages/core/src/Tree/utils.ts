export function flatten<U, T extends { children: Array<any> } >(items: Array<T>): Array<U> {
  return items.reduce((acc: Array<any>, item: T) => {
    acc.push(item);

    if (item.children) {
      acc.push(...flatten(item.children));
    }

    return acc;
  }, []);
}

// TODO: expose more utility function to handle flattened item
