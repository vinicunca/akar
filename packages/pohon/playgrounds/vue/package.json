{
  "name": "playground-vue",
  "type": "module",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "dev": "vite",
    "build": "vite build",
    "preview": "vite preview",
    "typecheck": "vue-tsc -p ./tsconfig.app.json"
  },
  "dependencies": {
    "pohon-ui": "workspace:*",
    "vue": "catalog:core",
    "vue-router": "catalog:base"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "catalog:core",
    "typescript": "catalog:base",
    "vite": "catalog:core",
    "vue-tsc": "catalog:core"
  }
}
