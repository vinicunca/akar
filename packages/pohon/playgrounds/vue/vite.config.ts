import vue from '@vitejs/plugin-vue';
import pohon from 'pohon/vite';
import { defineConfig } from 'vite';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    vue(),
    pohon({
      ui: {
        colors: {
          primary: 'green',
          neutral: 'slate',
        },
      },
      autoImport: {
        imports: ['vue'],
      },
      components: {
        dirs: ['../nuxt/app/components'],
      },
    }),
  ],
  optimizeDeps: {
    // prevents reloading page when navigating between components
    include: ['@unhead/vue/client', '@iconify/vue', '@internationalized/date', '@vueuse/shared', '@vueuse/integrations/useFuse', '@tanstack/vue-table', 'reka-ui', 'reka-ui/namespaced', 'embla-carousel-vue', 'embla-carousel-autoplay', 'embla-carousel-auto-scroll', 'embla-carousel-auto-height', 'embla-carousel-class-names', 'embla-carousel-fade', 'embla-carousel-wheel-gestures', 'colortranslator', 'tailwindcss/colors', 'tailwind-variants', 'ufo', 'zod', 'vaul-vue'],
  },
});
