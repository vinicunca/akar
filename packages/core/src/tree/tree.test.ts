import type { DOMWrapper, VueWrapper } from '@vue/test-utils';
import { KEY_CODES } from '@vinicunca/perkakas';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { nextTick } from 'vue';
import Tree from './story/_tree.vue';

describe('given default Tree', () => {
  // const kbd = useKbd()
  let wrapper: VueWrapper<InstanceType<typeof Tree>>;
  let content: DOMWrapper<Element>;
  let items: Array<DOMWrapper<Element>>;

  const updateItems = () => {
    items = wrapper.findAll('[role=treeitem]');
  };

  beforeEach(() => {
    document.body.innerHTML = '';
    wrapper = mount(Tree, { attachTo: document.body, props: { selectionBehavior: 'toggle' } });
    content = wrapper.find('[role=tree]');
    updateItems();
  });

  it('should pass axe accessibility tests', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations();
  });

  it('should render snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should select and deselect item', async () => {
    const item = items[0];
    await item.trigger('click');
    expect(item.attributes('aria-selected')).toBe('true');
    await item.trigger('click');
    expect(item.attributes('aria-selected')).toBe('false');
  });

  describe('when expand item by press ArrowRight', async () => {
    beforeEach(async () => {
      await items[1].trigger('keydown', { key: KEY_CODES.ARROW_RIGHT });
      (items[1].element as HTMLElement).focus();
      updateItems();
    });

    it('should pass axe accessibility tests', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });

    it('should expand the item, revealing it\'s item', () => {
      expect(items[2].text()).toBe('tree');
    });

    it('should close when press ArrowLeft', async () => {
      await items[1].trigger('keydown', { key: KEY_CODES.ARROW_LEFT });
      updateItems();
      expect(items[2].text()).toBe('routes');
      expect(items[2].text()).not.toBe('tree');
    });

    it('should focus on parent when press ArrowLeft on child item', async () => {
      await items[2].trigger('keydown', { key: KEY_CODES.ARROW_DOWN });
      await items[3].trigger('keydown', { key: KEY_CODES.ARROW_LEFT });
      expect(document.activeElement).toBe(items[1].element);
    });

    it('should focus on child item when press ArriwRight', async () => {
      await items[1].trigger('keydown', { key: KEY_CODES.ARROW_RIGHT });
      expect(document.activeElement).toBe(items[2].element);
    });

    describe('when expand nested item', async () => {
      beforeEach(async () => {
        await items[2].trigger('keydown', { key: KEY_CODES.ARROW_RIGHT });
        updateItems();
      });

      it('should pass axe accessibility tests', async () => {
        expect(await axe(wrapper.element)).toHaveNoViolations();
      });

      it('should expand the nested item, revealing it\'s item ', () => {
        expect(items[3].text()).toBe('Tree.vue');
      });
    });
  });

  // Test: useTypeAhead
  describe('when typing letter', async () => {
    it('should highlight text starting with l', async () => {
      await content.trigger('keydown', { key: 'l' });
      const item = items.find((i) => i.text().startsWith('l'));
      expect(document.activeElement).toBe(item?.element);
    });
  });

  describe('when selection behavior `replace`', () => {
    beforeEach(async () => {
      await wrapper.setProps({ selectionBehavior: 'replace' });
      await nextTick();
    });

    it('should not toggle off the selected value', async () => {
      const item = items[0];
      await item.trigger('click');
      await item.trigger('click');
      expect(item.attributes('aria-selected')).toBe('true');
    });

    it('should select and replace another item', async () => {
      const item = items[0];
      const newItem = items[1];
      await item.trigger('click');
      expect(item.attributes('aria-selected')).toBe('true');
      await newItem.trigger('click');
      expect(item.attributes('aria-selected')).toBe('false');
      expect(newItem.attributes('aria-selected')).toBe('true');
    });
  });
});

describe('given multiple `true` Tree', () => {
  let wrapper: VueWrapper<InstanceType<typeof Tree>>;
  // let content: DOMWrapper<Element>
  let items: Array<DOMWrapper<Element>>;

  beforeEach(async () => {
    document.body.innerHTML = '';
    wrapper = mount(Tree, { props: { multiple: true, selectionBehavior: 'toggle' }, attachTo: document.body });
    await nextTick();
    // content = wrapper.find('[role=tree]')
    items = wrapper.findAll('[role=treeitem]');
  });

  it('should select multiple items', async () => {
    await items[0].trigger('keydown', { key: KEY_CODES.ENTER });
    await items[0].trigger('keydown', { key: KEY_CODES.ARROW_DOWN });
    await items[1].trigger('keydown', { key: KEY_CODES.ARROW_DOWN });
    await items[2].trigger('keydown', { key: KEY_CODES.ENTER });

    expect(items[0].attributes('aria-selected')).toBe('true');
    expect(items[1].attributes('aria-selected')).toBe('false');
    expect(items[2].attributes('aria-selected')).toBe('true');
  });

  describe('when selection behavior `replace`', () => {
    beforeEach(async () => {
      wrapper.setProps({ selectionBehavior: 'replace' });
      await nextTick();
      await items[0].trigger('click');
      (items[0].element as HTMLElement).focus();
    });

    it('should not toggle off the selected value', async () => {
      const item = items[0];
      await item.trigger('click');
      await item.trigger('click');
      expect(item.attributes('aria-selected')).toBe('true');
    });

    it('should select and replace another item', async () => {
      const item = items[0];
      const newItem = items[1];
      expect(item.attributes('aria-selected')).toBe('true');
      await newItem.trigger('click');
      expect(item.attributes('aria-selected')).toBe('false');
      expect(newItem.attributes('aria-selected')).toBe('true');
    });

    describe('when keypress Shift + ArrowDown', () => {
      it('should select the next item', async () => {
        await items[0].trigger('keydown.shift', { key: KEY_CODES.ARROW_DOWN });
        expect(items[0].attributes('aria-selected')).toBe('true');
        expect(items[1].attributes('aria-selected')).toBe('true');
        expect(items[2].attributes('aria-selected')).toBe('false');

        await items[1].trigger('keydown.shift', { key: KEY_CODES.ARROW_DOWN });
        expect(items[0].attributes('aria-selected')).toBe('true');
        expect(items[1].attributes('aria-selected')).toBe('true');
        expect(items[2].attributes('aria-selected')).toBe('true');
      });
    });
  });
});

describe('given a Tree with a custom children structure', () => {
  const customItems = [
    { title: 'index.vue', icon: 'vue' },
    {
      title: 'lib',
      icon: 'folder',
      directories: [
        {
          title: 'tree',
          icon: 'folder',
          files: [
            {
              title: 'Tree.vue',
              icon: 'vue',
            },
            {
              title: 'TreeView.vue',
              icon: 'vue',
            },
          ],
        },
        {
          title: 'icons',
          icon: 'folder',
          files: [
            { title: 'JS.vue', icon: 'vue' },
            { title: 'vue.vue', icon: 'vue' },
          ],
        },
      ],
      files: [
        {
          title: 'index.js',
          icon: 'js',
        },
      ],
    },
    {
      title: 'routes',
      icon: 'folder',
      directories: [
        {
          title: 'contents',
          icon: 'folder',
          files: [
            {
              title: '+layout.vue',
              icon: 'vue',
            },
            {
              title: '+page.vue',
              icon: 'vue',
            },
          ],
        },
      ],
    },
  ];

  let wrapper: VueWrapper<InstanceType<typeof Tree>>;
  // let content: DOMWrapper<Element>
  let items: Array<DOMWrapper<Element>>;

  beforeEach(async () => {
    document.body.innerHTML = '';
    wrapper = mount(Tree, {
      props: {
        items: customItems,
        multiple: true,
        selectionBehavior: 'toggle',
        getChildren: (val) => {
          if (!val.files) {
            return val.directories;
          }

          if (!val.directories) {
            return val.files;
          }

          return [...val.directories, ...val.files]; ;
        },
      },
      attachTo: document.body,
    });
    await nextTick();
    // content = wrapper.find('[role=tree]')
    items = wrapper.findAll('[role=treeitem]');
  });

  const updateItems = () => {
    items = wrapper.findAll('[role=treeitem]');
  };

  describe('when expand item', async () => {
    beforeEach(async () => {
      await items[1].trigger('keydown', { key: KEY_CODES.ARROW_RIGHT });
      (items[1].element as HTMLElement).focus();
      updateItems();
    });

    it('should expand the item, revealing it\'s item', () => {
      expect(items[2].text()).toBe('tree');
    });

    describe('when expand nested item', async () => {
      beforeEach(async () => {
        await items[2].trigger('keydown', { key: KEY_CODES.ARROW_RIGHT });
        updateItems();
      });

      it('should expand the nested item, revealing it\'s item ', () => {
        expect(items[3].text()).toBe('Tree.vue');
      });
    });
  });
});

describe('given a Tree with bubbleSelect and propagateSelect', () => {
  const customItems = [
    {
      title: 'components',
      children: [
        {
          title: 'Home',
          children: [
            { title: 'Card.vue' },
            { title: 'Button.vue' },
          ],
        },
      ],
    },
  ];

  let wrapper: VueWrapper<InstanceType<typeof Tree>>;
  let items: Array<DOMWrapper<Element>>;

  beforeEach(async () => {
    document.body.innerHTML = '';
    wrapper = mount(Tree, { props: { items: customItems, expanded: ['components', 'Home'], multiple: true, propagateSelect: true, bubbleSelect: true } });
    await nextTick();
    items = wrapper.findAll('[role=treeitem]');
  });

  it('propagates changes down the tree', async () => {
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['false', 'false', 'false', 'false']);
    await items[0].trigger('click');
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['true', 'true', 'true', 'true']);
    await items[0].trigger('click');
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['false', 'false', 'false', 'false']);
  });

  it('bubbles change up the tree', async () => {
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['false', 'false', 'false', 'false']);
    await items[2].trigger('click');
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['false', 'false', 'true', 'false']);
    await items[3].trigger('click');
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['true', 'true', 'true', 'true']);
    await items[2].trigger('click');
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['false', 'false', 'false', 'true']);
    await items[3].trigger('click');
    expect(items.map((i) => i.attributes('aria-selected'))).toStrictEqual(['false', 'false', 'false', 'false']);
  });
});
