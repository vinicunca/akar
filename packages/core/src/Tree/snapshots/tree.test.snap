// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given default Tree > should render snapshot 1`] = `
"<ul role="tree" tabindex="0" data-orientation="vertical" dir="ltr" style="outline: none;" class="list-none select-none w-64 bg-white text-blackA11 rounded-md p-2 text-sm font-medium">
  <li class="flex items-center py-1 px-2 my-0.5 rounded w-max outline-none focus:ring-grass9 focus:ring-2 data-[selected]:bg-grass4" style="margin-left: 1rem;" aria-setsize="3" aria-posinset="1" role="treeitem" aria-selected="false" aria-level="1" data-indent="1" tabindex="-1" data-orientation="vertical" data-active="true" data-akar-collection-item="">
    <!--v-if-->
    <div class="pl-2">index.vue</div>
  </li>
  <li class="flex items-center py-1 px-2 my-0.5 rounded w-max outline-none focus:ring-grass9 focus:ring-2 data-[selected]:bg-grass4" style="margin-left: 1rem;" aria-setsize="3" aria-posinset="2" role="treeitem" aria-selected="false" aria-expanded="false" aria-level="1" data-indent="1" tabindex="-1" data-orientation="vertical" data-active="true" data-akar-collection-item=""><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" role="img" class="h-4 w-4" width="1em" height="1em" viewBox="0 0 16 16"></svg>
    <div class="pl-2">lib</div>
  </li>
  <li class="flex items-center py-1 px-2 my-0.5 rounded w-max outline-none focus:ring-grass9 focus:ring-2 data-[selected]:bg-grass4" style="margin-left: 1rem;" aria-setsize="3" aria-posinset="3" role="treeitem" aria-selected="false" aria-expanded="false" aria-level="1" data-indent="1" tabindex="-1" data-orientation="vertical" data-active="true" data-akar-collection-item=""><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" role="img" class="h-4 w-4" width="1em" height="1em" viewBox="0 0 16 16"></svg>
    <div class="pl-2">routes</div>
  </li>
</ul>"
`;
